text/0314-cdk-construct-hub.md

```diff
@@ -0,0 +1,626 @@
+  - API reference documentation, in all supported languages (only the version
+    for the currently selected language is displayed)
+
+The list of available languages is also a selector: upon clicking a language,
+the installation instructions and documentation automatically switch to the
+relevant form for the currently selected language.
+
+### Support
+
+The Construct Hub offers ways for customers to engage maintainers about
+various problems, including the following (actually routed to GitHub issue
+templates):
+
+- A malicious package was identified
+- Incorrectly rendered documentation
+- Package un-listing request
+
+When a contact request originates from a package detail page, the customer is
+guided to the package's issue tracker or public repository if their issue is
+with the package itself (and not about it's listing in Construct Hub). The
+routing is based on metadata configured on the package itself, through the bug
+tracker URL, repository URL, and homepage URL.
+
+## Technical Design
+
+### Overview
+
+![Architecture Diagram](../images/cdk-construct-hub/architecture-diagram.png)
+
+### What is the high level implementation plan?
+
+#### Overview
+
+The Construct Hub is a relatively simple static web application. While the
+dataset it exposes (all Construct Libraries) may be relatively fast moving,
+there is no necessity for newly published packages (or versions thereof) to be
+indexed and presented on the application particularly quickly. A consistency
+window between a couple of hours and a day is probably acceptable.
+
+A simple data pipeline will ingest package publication events created by a
+source-specific adapter (e.g: by polling on the CouchDB instance replica for the
+[npmjs.com registry](https://skimdb.npmjs.com/registry)), and produce static
+artifacts the website is based off of. This pipeline is in particular
+responsible for extracting the `.jsii` assembly file from the npm packages, and
+to prefetch all information necessary to back the webiste's search and package
+detail pages.
+
+#### Back-End
+
+1. **Package Discovery:** The back-end for Construct Hub starts with a custom
+   built event source, which implements the logic necessary to detect new
+   packages that are relevant to the Construct Hub, and sends messages to an SQS
+   queue for further processing. Notifications have the following attributes:
+
+   Name        | Description
+   ------------|----------------------------------------------------------------
+   `assembly`  | The `.jsii` assembly included in the package
+   `time`      | The timestamp at which the version was created
+   `integrity` | An integirity check for the complete record
+
+   In cases where the `assembly` object is too large, it may be staged in S3
+   instead of being sent directly as part of the message payload. In such cases,
+   the `assembly` field contains an S3 URI instead of a JSON object.
+
+   Construct Hub comes with a built-in integration function that listents to
```

Comment:
I am assuming we are planning to simply query npm and filter versions we've already seen, right?

```suggestion
   Construct Hub comes with a built-in integration function that listens to
```

Reply:
What's the protocol of the package source? In other words, what would I need to do in order to implement my own source? Do I just need to push messages to a queue? What do these messages contain?

Reply:
> I am assuming we are planning to simply query npm and filter versions we've already seen, right?

No, we are going to query the CouchDB replica's change stream, to discover newly published package versions (and only those).

---

> In other words, what would I need to do in order to implement my own source? Do I just need to push messages to a queue? What do these messages contain?

Yes, you publish to SQS, and the contents of the messages is what is described/specified in the relevant section of this document.

Reply:
> we are going to query the CouchDB replica's change stream, to discover newly published package versions (and only those).

Can you describe the protocol in the doc?  Does this mean we need to store our cursor state somewhere? What's the plan for backfilling existing data?